### PR TITLE
Fix cache warming rapid loop

### LINKED ISSUE #926: Flux using >50% of one CPU
Noticing multiple `kubectl apply` every second.

This occurs with v1.2.0 and v1.2.1.

### PR DESCRIPTION
**Pull request by stefanprodan**
Fix #926

Tested with `stefanprodan/flux:cpufix-090b064-wip3` on GKE:

screen shot 2018-02-03 at 16 46 05

### CODE REVIEW

**Comment by rade** [comment]
FWIW, the spin loop was introduced in #835.

File: registry/cache/warming.go [hunk 1]
```diff
@@ -91,10 +91,18 @@ func (w *Warmer) Loop(logger log.Logger, stop <-chan struct{}, wg *sync.WaitGrou
 			w.warm(ctx, logger, im.Name, im.Credentials)
 		} else {
 			select {
+			case <-stop:
+				logger.Log("stopping", "true")
```
Comment (rade) [inline registry/cache/warming.go]: there's a `return` missing here.

File: registry/cache/warming.go [hunk 2]
```diff
@@ -91,10 +91,18 @@ func (w *Warmer) Loop(logger log.Logger, stop <-chan struct{}, wg *sync.WaitGrou
...
 				imageCreds = imagesToFetchFunc()
 				backlog = imageCredsToBacklog(imageCreds)
-			default:
+			case name := <-w.Priority:
+				logger.Log("priority", name.String())
+				if creds, ok := imageCreds[name]; ok {
+					w.warm(ctx, logger, name, creds)
+				} else {
+					logger.Log("priority", name.String(), "err", "no creds available")
+				}
```
Comment (rade) [inline registry/cache/warming.go]: I'd be tempted to lift these six lines and their copies (plus comment) in the earlier `select` into a
```golang
// ...
priorityWarm := func (name string) {
  ...
}
```
before the loop.

**Comment by rade** [comment]
Why is this not consuming an entire core all the time? Is that because every `askForNewImagesInterval` (1min) it does some real work, which involves i/o and hence releases the CPU?

**Comment by stefanprodan** [comment]
Yes every time the refresh timer hits it does some work for a couple of seconds. The rest of the time just sits in the default loop.

screen shot 2018-02-03 at 20 10 45

**Comment by squaremo** [comment]
Looks good in sum @stefanprodan, thank you. Thanks @rade for helping knock it into shape.

Please rebase into a single commit -- or two, if you want to do the factoring of `priorityWarm` separately -- and write a short commit message explaining what the problem was and how this fixes it.

**Comment by stefanprodan** [comment]
@squaremo I've squashed the commits into one. 
@rade thanks for guiding me through this.

**Comment by stefanprodan** [comment]
@squaremo let me know if it's ok for me to merge this PR.

**Comment by squaremo** [comment]
> I've squashed the commits into one.

Thank you. I am a bit baffled by the commit message though -- did you really add those things? (I don't think you did.) Something like this would describe what's being fixed:

> To implement the loop as described in the comment above it, when the backlog is empty we need to block until the refresh tick fires (or a priority image comes in). By having a default case, it's possible to spin.

(the commit summary can be a bit shorter too -- just "Avoid spin in cache refresh loop" would be fine IMO).

**Pull request merged**